R/init.R

```diff
@@ -122,52 +121,59 @@ init <- function(data,
       )
     )
   }
-
+  if (is.list(data) && !inherits(data, "teal_data_module")) {
+    checkmate::assert_list(data, names = "named")
+    data <- do.call(teal.data::teal_data, data)
+  }
   checkmate::assert_multi_class(data, c("teal_data", "teal_data_module"))
-  checkmate::assert_multi_class(modules, c("teal_module", "list", "teal_modules"))
+
+  ## `modules`
+  if (inherits(modules, "teal_module")) {
+    modules <- list(modules)
+  }
+  if (inherits(modules, "list")) {
+    modules <- do.call(teal::modules, modules)
+  }
+  checkmate::assert_class(modules, "teal_modules")
```

Comment:
> Changing checkmate::assert_multi_class(modules, c("teal_module", "list", "teal_modules")) pre-normalization to checkmate::assert_class(modules, "teal_modules") post-normalization removes the gate on allowed argument types.

Hmm I don't follow. Could you please elaborate more?

My rationale to do convert - assert (as opposed to assert - convert) is that we can also include the results of convert in the assertion that follows (note that we dont do assert on return inside convert).

> Additionally, if modules is list, it wouldn't hurt to check if elements are teal_module(s).

I was thinking about this and this is already covered in here: 
https://github.com/insightsengineering/teal/blob/few_enhancements/R/modules.R#L78
```
checkmate::assert_list(submodules, min.len = 1, any.missing = FALSE, types = c("teal_module", "teal_modules"))
```

Reply:
> > Changing checkmate::assert_multi_class(modules, c("teal_module", "list", "teal_modules")) pre-normalization to checkmate::assert_class(modules, "teal_modules") post-normalization removes the gate on allowed argument types.
> 
> Hmm I don't follow. Could you please elaborate more?

I prefer checking what is passed into the function so that we can fail specifically rather than attempt a conversion from anything and then fail on the attempt. One could pass a `FilteredData` as modules (because why not?) and `do.call(teal::modules, modules)` would fail miserably.

Reply:
> rather than attempt a conversion from anything and then fail on the attempt

We don't try on anything - note that each conversion is conditional. The goal is to have the following:

```
if (inherits(x, "a")) convert_a_to_x(x)
if (inherits(x, "b")) convert_b_to_x(x)
assert_class(x, "x")
```

But you have discovered one thing - R6 objects inherit from list and `(inherits(modules, "list"))` condition is imperfect. How about changing this into `if (checkmate::test_list(modules, types = c("teal_module", "teal_modules"))) {`?

> One could pass a FilteredData as modules (because why not?) and do.call(teal::modules, modules) would fail miserably.

Great example and love this way of thinking. The goal is that this won't enter any conversion and fall directly into the assertion. That's not the case currently but will be true after proposal.

Reply:
I favor the `assert ➟ convert` workflow when we can group the assertions on the same argument together (instead of having 2+ assertions as `data` has).

If we can protect against all allowable types, then we can "safely" convert those. Although I see a point in making sure that in the end, we have what is expected.

<details>

<summary>Sample</summary>

```r
## `data`
checkmate::assert(
  combine = "or",
  .var.name = "data",
  checkmate::check_multi_class(data, c("teal_data", "teal_data_module")),
  checkmate::check_list(data, names = "named")
)

# then convert

## `modules`

checkmate::assert(
  combine = "or",
  .var.name = "modules",
  checkmate::check_multi_class(modules, c("teal_module", "teal_modules")),
  checkmate::check_list(modules, types = c("teal_module", "teal_modules"))
)

# then convert
```

</details>

> But you have discovered one thing - R6 objects inherit from list and (inherits(modules, "list")) condition is imperfect

``` r
Sample <- R6::R6Class("Sample", public = list(sample = function(...) sample(...)))
inherits(Sample$new(), "list")
#> [1] FALSE

fd <- teal.slice:::FilteredData$new(list(iris = iris))
inherits(fd, "list")
#> [1] FALSE
```

Reply:
Right - true. I was wrong with my statement that R6 instances inherits from list. Then the example of FilteredData as data should behave correctly and no changes needed.

Reply:
@pawelru 
```
 if (inherits(modules, "list")) {
    modules <- do.call(teal::modules, modules)
  }
```
All right, a **list** of `FilteredData`s.

@averissimo 
> If we can protect against all allowable types, then we can "safely" convert those. 

Exactly. :+1: 

> Although I see a point in making sure that in the end, we have what is expected.

Our conversion routines/procedures depend _only_ on the input, so as long their logic is sound, checking results seems redundant to me. Of course, we are open to the problem of shifting definitions, e.g. we use a `foo` to convert `bar` to `pub` but then then `bar` stops being a list and becomes an S4 object and `foo` fails. But we have unit tests for such cases, do we not?

Reply:
Hey guys. Thanks for all of this. It seems that there is an agreement among developers so I implemented this in a suggested way. There are a few additional small caveats that I have discovered down the way but we can also live with that I guess.

Reply:
(resolving this thread)